pkg/storage/allocator_test.go

```diff
@@ -685,9 +694,164 @@ func TestAllocatorRebalance(t *testing.T) {
+	defer stopper.Stop(context.Background())
+	stores := []*roachpb.StoreDescriptor{
+		{
+			StoreID: 1,
+			Node: roachpb.NodeDescriptor{
+				NodeID: 1,
+				Locality: roachpb.Locality{
+					Tiers: []roachpb.Tier{
+						{Key: "datacenter", Value: "a"},
+					},
+				},
+			},
+			Capacity: roachpb.StoreCapacity{
+				Capacity:        100,
+				Available:       80,
+				RangeCount:      50,
+				LogicalBytes:    50,
+				WritesPerSecond: 30,
+			},
+		},
+		{
+			StoreID: 2,
+			Node: roachpb.NodeDescriptor{
+				NodeID: 2,
+				Locality: roachpb.Locality{
+					Tiers: []roachpb.Tier{
+						{Key: "datacenter", Value: "a"},
+					},
+				},
+			},
+			Capacity: roachpb.StoreCapacity{
+				Capacity:        100,
+				Available:       80,
+				RangeCount:      50,
+				LogicalBytes:    50,
+				WritesPerSecond: 40,
+			},
+		},
+		{
+			StoreID: 3,
+			Node: roachpb.NodeDescriptor{
+				NodeID: 3,
+				Locality: roachpb.Locality{
+					Tiers: []roachpb.Tier{
+						{Key: "datacenter", Value: "b"},
+					},
+				},
+			},
+			Capacity: roachpb.StoreCapacity{
+				Capacity:        100,
+				Available:       80,
+				RangeCount:      55,
+				LogicalBytes:    50,
+				WritesPerSecond: 30,
+			},
+		},
+		{
+			StoreID: 4,
+			Node: roachpb.NodeDescriptor{
+				NodeID: 4,
+				Locality: roachpb.Locality{
+					Tiers: []roachpb.Tier{
+						{Key: "datacenter", Value: "c"},
+					},
+				},
+			},
+			Capacity: roachpb.StoreCapacity{
+				Capacity:        100,
+				Available:       30,
+				RangeCount:      150,
+				LogicalBytes:    80,
+				WritesPerSecond: 100,
+			},
+		},
+		{
+			StoreID: 5,
+			Node: roachpb.NodeDescriptor{
+				NodeID: 5,
+				Locality: roachpb.Locality{
+					Tiers: []roachpb.Tier{
+						{Key: "datacenter", Value: "d"},
+					},
+				},
+			},
+			Capacity: roachpb.StoreCapacity{
+				Capacity:        100,
+				Available:       50,
+				RangeCount:      70,
+				LogicalBytes:    70,
+				WritesPerSecond: 60,
+			},
+		},
+	}
+	sg := gossiputil.NewStoreGossiper(g)
+	sg.GossipStores(stores, t)
+
+	st := a.storePool.st
+	EnableStatsBasedRebalancing.Override(&st.SV, true)
+	replicas := []roachpb.ReplicaDescriptor{
+		{
+			StoreID: 1,
+			NodeID:  1,
+		},
+		{
+			StoreID: 3,
+			NodeID:  3,
+		},
+		{
+			StoreID: 4,
+			NodeID:  4,
+		},
+	}
+	repl := &Replica{RangeID: firstRange}
+
+	repl.mu.Lock()
+	repl.mu.state.Stats = &enginepb.MVCCStats{}
+	repl.mu.Unlock()
+
+	rs := newReplicaStats(clock, nil)
+	repl.writeStats = rs
+
+	status := &raft.Status{
+		Progress: make(map[uint64]raft.Progress),
+	}
+	for _, replica := range replicas {
+		status.Progress[uint64(replica.NodeID)] = raft.Progress{
+			Match: 10,
+		}
+	}
+	for i := 0; i < 10; i++ {
+		result, _ := a.RebalanceTarget(
+			context.Background(),
+			config.Constraints{},
+			repl,
+			status,
+			testRangeInfo(replicas, firstRange),
+			storeFilterThrottled,
+		)
+		if expected := roachpb.StoreID(5); result.StoreID != expected {
```

Comment:
I don't think having a fifth store here is any better than just having 4 stores and verifying that we don't want to move a replica to the fourth.

Reply:
Yeah, you are right.

Reply:
Yeah, you are right.  4 stores will be enough.